src/frontend/SignatureMismatch.ml

```diff
@@ -170,34 +189,62 @@ let promote es promotions =
       let open UnsizedType in
       let emeta = exp.emeta in
       match prom with
-      | RealPromotion when is_int_type emeta.type_ ->
+      | IntToRealPromotion when is_int_type emeta.type_ ->
           Ast.
             { expr= Ast.Promotion (exp, UReal, emeta.ad_level)
             ; emeta= {emeta with type_= promote_array emeta.type_ UReal} }
-      | ComplexPromotion when not (is_complex_type emeta.type_) ->
+      | (IntToComplexPromotion | RealToComplexPromotion)
+        when not (is_complex_type emeta.type_) ->
           { expr= Promotion (exp, UComplex, emeta.ad_level)
           ; emeta= {emeta with type_= promote_array emeta.type_ UComplex} }
       | _ -> exp )
 
-let returntype env name args =
+let promotion_cost p =
+  match p with
+  | None -> 0
+  | RealToComplexPromotion | IntToRealPromotion -> 1
+  | IntToComplexPromotion -> 2
+
+let unique_minimum_promotion ps =
```

Comment:
Why is the name of this input `ps`?

Reply:
`ps` is the plural of `p` which is short for `promotion`. That's my guess at least.

Reply:
Niko is correct on the logic. You run into fun plural-of-a-plural issues with trying to describe a list of promotion*s*, but `ps` worked.

I tend to abbreviate non-globals or variables whose meaning should be clear from their type. I suppose it's the same reason most people don't write `for(int index = 0; ...` . I can do a pass on the really short names to try to clear some of them up

Reply:
> I tend to abbreviate non-globals or variables whose meaning should be clear from their type. I suppose it's the same reason most people don't write for(int index = 0; ... . I can do a pass on the really short names to try to clear some of them up

ftr it took me a minute to long to realize what `ps` is and what it's purpose was. idt we should rely heavily on the ocaml compilers type deduction in order for us to understand code. (I've seen code @nhuurre has fixed for me that for some reason needed type hinting to compile!). If we used type hinting everywhere I'd be cool with smaller names but we don't so I'd prefer more meaningful names for anything that is not trivial like an iterator.

> I tend to abbreviate non-globals or variables whose meaning should be clear from their type. I suppose it's the same reason most people don't write for(int index = 0;

I have a lot of thoughts about this sentence but I would rather keep my comment short and say idt the meaning or intent is clear from the type given by the Ocaml Compiler™ and we should either have meaningful names, type hints, or both. imo I'd also prefer at least one liner docs here briefly describing what functions we make do.

Reply:
Are you trying to suggest ` ('a * promotion list) list -> ('a * promotion list, 'a list option) result` _isn't_ a crystal clear type?
(Obviously joking here)

I think if you spend all day reading OCaml that can start to look sensible, in the same way C++ templating magic comes naturally to people who do it a lot, and I'm probably guilty of just being too familiar with my own code.  I've changed the name to `promotion_candidates`. 

To elaborate a bit more on the original name, I think it's common for lists to be given some very short name ending in s. This is hardly a perfect convention, but I tend to read that as meaning that the fact that it _is_ a list is more important than what it is a list _of_ (whether or not this applies to this function is highly debatable). 

Anyway, when I wrote the first version of the function (which was originally only used one place, so it wasn't even part of the module at the top level) I just picked something short. When I needed to do the same thing somewhere else, I promoted the function without thinking about renaming it's argument.